File: fizzbuzz/fizzbuzz.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

#define DEFAULT_FIZZBUZZ_COUNT  100

#define FL_DIV_NONE     0
#define FL_DIV_BY_THREE (1 << 0)
#define FL_DIV_BY_FIVE  (1 << 1)

#define STR_DIV_THREE_MSG   "Fizz"
#define STR_DIV_FIVE_MSG    "Buzz"

#define STR_DIV_SPACE   1

#ifndef TRUE
#   define TRUE 1
#endif

#ifndef FALSE
#   define FALSE 0
#endif

#define IVN_STATE_SIGN        0
#define IVN_STATE_DIGIT       1
#define IVN_STATE_EXIT        2

int is_valid_number(const char* str)
{
    const char* pStr = str;

    int result = 
        str && *str != '\0' ? TRUE : FALSE;

    int state = IVN_STATE_SIGN;

    int numDigits = 0;

    if (result)
    {
        while (state != IVN_STATE_EXIT && 
            *pStr != '\0')
        {
            switch (state)
            {
                case IVN_STATE_SIGN:
                    if (*pStr == '-' || 
                        *pStr == '+' ||
                        isdigit(*pStr))
                    {
                        state = IVN_STATE_DIGIT;

                        if (isdigit(*pStr))
                            pStr--;
                    }
                    else
                    {
                        result = FALSE;
                        state = IVN_STATE_EXIT;
                    }
                break;
                case IVN_STATE_DIGIT:
                    if (!isdigit(*pStr))
                    {
                        result = FALSE;
                        state = IVN_STATE_EXIT;
                    }
                    else
                    {
                        numDigits++;
                    }
                break;
            }

            pStr++;
        }

        if (numDigits < 1)
            result = FALSE;
    }

    return result;
}

int main(int argc, const char** argv)
{
    int fzCount = DEFAULT_FIZZBUZZ_COUNT;
    int fzFlags = FL_DIV_NONE;

    int result = 0;

    int i, ai;

    if (argc > 1)
    {
        for (ai = 1; ai < argc; ai++)
        {
            if (is_valid_number(argv[ai]))
            {
                fzCount = strtol(argv[ai], NULL, 10);
            }
            else
            {
                fprintf(stderr, "Invalid argument: '%s'\n", argv[ai]);
                result = 1;
            }
        }
    }

    if (fzCount < 1)
    {
        fprintf(stderr, "Number to count up to must be greater than or equal to one.\n");
        result = 1;
    }

    if (result == 0)
    {
        printf("Counting up to %i:\n", fzCount);

        for (i = 1; i <= fzCount; i++)
        {
            if (!(i % 3))
                fzFlags |= FL_DIV_BY_THREE;
            
            if (!(i % 5))
                fzFlags |= FL_DIV_BY_FIVE;

            if (fzFlags)
            {
                if (fzFlags & FL_DIV_BY_THREE)
                    printf(STR_DIV_THREE_MSG);

#if STR_DIV_SPACE
                if ((fzFlags & FL_DIV_BY_THREE) &&
                    (fzFlags & FL_DIV_BY_FIVE))
                    printf(" ");
#endif

                if (fzFlags & FL_DIV_BY_FIVE)
                    printf(STR_DIV_FIVE_MSG);
            }
            else
            {
                printf("%i", i);
            }

            printf("\n");

            fzFlags = FL_DIV_NONE;
        }
    }

    return result;
}

File: fizzbuzz/fizzbuzz.js
#!/usr/bin/env node

const FL_DIVIDE_NONE = 0;
const FL_DIVIDE_BY_THREE = 1;
const FL_DIVIDE_BY_FIVE = 2;

const MSG_DIVIDE_BY_THREE = "Fizz";
const MSG_DIVIDE_BY_FIVE = "Buzz";
const MSG_DIVIDE_SPACE = " ";

const DEFAULT_FIZZBUZZ_COUNT = 100;

process.exitCode = (function (args) {
    let result = 0;
    let fizzBuzzCount = DEFAULT_FIZZBUZZ_COUNT;

    const writeOut = (str) => process.stdout.write(str);
    const writeErr = (str) => process.stderr.write(str);

    if (args !== null && args.length > 0)
    {
        for (let i = 0; i < args.length; i++)
        {
            if (isNaN(args[i]))
            {
                writeErr("Invalid argument: '" + args[i] + "'\n");
                result = 1;
            }
            else
            {
                fizzBuzzCount = args[i];
            }
        }
    }

    if (fizzBuzzCount < 1)
    {
        writeErr("Number to count up to must be greater than or equal to one.\n");
        result = 1;
    }

    if (result === 0)
    {
        writeOut("Counting up to " + fizzBuzzCount + ":\n");

        for (let i = 1; i <= fizzBuzzCount; i++)
        {
            let fzFlag = FL_DIVIDE_NONE;

            if ((i % 3) == 0)
                fzFlag |= FL_DIVIDE_BY_THREE;
            
            if ((i % 5) == 0)
                fzFlag |= FL_DIVIDE_BY_FIVE;
            
            if (fzFlag != FL_DIVIDE_NONE)
            {
                if ((fzFlag & FL_DIVIDE_BY_THREE) != 0)
                    writeOut(MSG_DIVIDE_BY_THREE);

                if ((fzFlag & (FL_DIVIDE_BY_THREE)) != 0 &&
                    (fzFlag & (FL_DIVIDE_BY_FIVE)) != 0)
                    writeOut(MSG_DIVIDE_SPACE);
                
                if ((fzFlag & FL_DIVIDE_BY_FIVE) != 0)
                    writeOut(MSG_DIVIDE_BY_FIVE);

                writeOut("\n");
            }
            else
            {
                writeOut(i + "\n");
            }
        }
    }

    return result;
})(process.argv.slice(2));